File: 2023 yard hoof cup/C++ The end scene/4.cpp
#include <bits/stdc++.h>

using namespace std;
#define int long long

const int mod = 337077;
int p;
int q;
int k;
int T;
int res;
vector<vector<bool>> mark;
const vector<vector<int>> dirs = {
    {0, 0},
    {1, 0},
    {-1, 0},
    {0, 1},
    {0, -1},
};

bool check(int x, int y) {
    for (const auto&dir: dirs) {
        int a = x + dir[0];
        int b = y + dir[1];
        if (a >= 0 && a < p && b >= 0 && b < q && mark[a][b]) {
            return false;
        }
    }
    return true;
}

void dfs(int pos, int man) {
    if (man == k) {
        res++;
        res %= mod;
        return;
    }
    for (int next = pos; next < p * q; next++) {
        int i = next / q;
        int j = next % q;
        if (check(i, j)) {
            mark[i][j] = true;
            dfs(next + 1, man + 1);
            mark[i][j] = false;
        }
    }
}

void solve() {
    cin >> T;
    ostringstream oss;
    for (int i = 0; i < T; i++) {
        cin >> p >> q >> k;
        if (k == 1) {
            oss << p * q << '\n';
        } else {
            mark.assign(p, vector<bool>(q, false));
            res = 0;
            dfs(0, 0);
            oss << res << '\n';
        }
    }
    cout << oss.str();
}

auto main() -> signed {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    solve();
    return 0;
}

File: 2023 yard hoof cup/Python The end scene/4.py
mod = 337077
dirs = [(0, 0), (1, 0), (-1, 0), (0, -1), (0, 1)]


def check(i: int, j: int) -> bool:
    for dir in dirs:
        a = i + dir[0]
        b = j + dir[1]
        if a >= 0 and a < p and b >= 0 and b < q and mark[a][b]:
            return False
    return True


def dfs(pos: int, man: int) -> int:
    if man == k:
        return 1
    res = 0
    for next in range(pos, p * q):
        i = next // q
        j = next % q
        if check(i, j):
            mark[i][j] = True
            res += dfs(next + 1, man + 1)
            res %= mod
            mark[i][j] = False
    return res


def solve():
    global p, q, T, k, mark
    T = int(input())
    ans = ""
    for i in range(T):
        p, q, k = map(int, input().split())
        if k == 1:
            ans += str(p * q) + '\n'
        else:
            mark = [[False for _ in range(q)] for _ in range(p)]
            ans += str(dfs(0, 0)) + '\n'
    print(ans, end='')


if __name__ == '__main__':
    solve()
